### PR TITLE
[Merged by Bors] - feat(Data/*/Sqrt): decidability of `IsSquare`

### PR DESCRIPTION
**Pull request by eric-wieser**
[Zulip thread](https://leanprover.zulipchat.com/#narrow/stream/217875-Is-there-code-for-X.3F/topic/.E2.9C.94.20Case.20checking.20in.20combinatorics/near/398517440)

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Algebra/Parity.lean [hunk 1]
```diff
@@ -63,8 +63,40 @@ theorem isSquare_op_iff (a : α) : IsSquare (op a) ↔ IsSquare a :=
 #align is_square_op_iff isSquare_op_iff
 #align even_op_iff even_op_iff
 
+@[to_additive]
+theorem isSquare_unop_iff {a : αᵐᵒᵖ} : IsSquare (unop a) ↔ IsSquare a := by
+  rw [←isSquare_op_iff, op_unop]
+
+@[to_additive]
+instance [DecidablePred (@IsSquare α _)] : DecidablePred (@IsSquare (αᵐᵒᵖ) _) :=
```
Comment (YaelDillies) [inline Mathlib/Algebra/Parity.lean:71]:
```suggestion
instance [@DecidablePred α IsSquare] : @DecidablePred αᵐᵒᵖ IsSquare :=
```

File: Mathlib/Data/Int/Sqrt.lean [hunk 2]
```diff
@@ -37,4 +37,9 @@ theorem sqrt_nonneg (n : ℤ) : 0 ≤ sqrt n :=
   coe_nat_nonneg _
 #align int.sqrt_nonneg Int.sqrt_nonneg
 
+/-- `IsSquare` can be decided on `ℤ` by checking against the square root.-/
```
Comment (YaelDillies) [inline Mathlib/Data/Int/Sqrt.lean:40]:
```suggestion
/-- `IsSquare` can be decided on `ℤ` by checking against the square root. -/
```

**Review by YaelDillies — COMMENTED** [review]
Style comments. The rest looks good!

File: Mathlib/Algebra/Parity.lean (hunk 1 above)
Comment (Vierkantor) [reply Mathlib/Algebra/Parity.lean:71]: What is the reason to prefer one over the other?

IMO we should be using the fancy new Lean 4 syntax:
```suggestion
instance [DecidablePred (α := α) IsSquare] : DecidablePred (α := αᵐᵒᵖ) IsSquare :=
```

**Review by Vierkantor — APPROVED** [review]
LGTM otherwise.

bors d=@YaelDillies

File: Mathlib/Algebra/Parity.lean (hunk 1 above)
Comment (eric-wieser) [reply Mathlib/Algebra/Parity.lean]: I went for a third option without `@` or specifying an argument name.

File: Mathlib/Algebra/Parity.lean [hunk 3]
```diff
@@ -63,8 +63,40 @@ theorem isSquare_op_iff (a : α) : IsSquare (op a) ↔ IsSquare a :=
 #align is_square_op_iff isSquare_op_iff
 #align even_op_iff even_op_iff
 
+@[to_additive]
+theorem isSquare_unop_iff {a : αᵐᵒᵖ} : IsSquare (unop a) ↔ IsSquare a := by
+  rw [←isSquare_op_iff, op_unop]
```
Comment (YaelDillies) [inline Mathlib/Algebra/Parity.lean:68]:
```suggestion
theorem isSquare_unop_iff {a : αᵐᵒᵖ} : IsSquare (unop a) ↔ IsSquare a := isSquare_op_iff.symm
```
no?

Comment (eric-wieser) [reply Mathlib/Algebra/Parity.lean]: Done

**Review by YaelDillies — APPROVED** [review]
Nicely done!

bors merge

**Pull request closed without merging**